Google Tag Manager Connector not adding GTM script

## Description of bug
When activating the Google Tag Manager Connector the GTM script is not added to the head section. Instead the logs show a null reference exception. Only after replacing root/SiteAnalytics.config by the content of root/DesktopModules/Connectors/GoogleTagManager/GoogleTagManager.config, it's working as intended.


## Steps to reproduce
List the precise steps to reproduce the bug:
1. Change Settings -> Connectors from Google Analytics to Google Tag Manager
2. Clear cache, restart server
3. Clear browser cache and reload
4. Check page source and search for Google Tag Manager script

## Current behavior
GTM Connector not adding code to <head> section

## Expected behavior
GTM Connector adding gtm script to <head> section

## Screenshots
If applicable, provide screenshots to help explain the bug.

## Error information
[D:12][T:121][ERROR] DotNetNuke.Services.Analytics.Config.AnalyticsConfiguration - System.NullReferenceException: Object reference not set to an instance of an object.
   at DotNetNuke.Services.Analytics.Config.AnalyticsConfiguration.GetConfig(String analyticsEngineName) in D:\a\1\s\DNN Platform\Library\Services\Analytics\Config\AnalyticsConfiguration.cs:line 87

See https://github.com/dnnsoftware/Dnn.Platform/blob/399a89c292cc8fe1a3ce66a7c592f15104bfbaae/DNN%20Platform/Library/Services/Analytics/Config/AnalyticsConfiguration.cs#L87

## Additional context
Possibly related to differences/missing settings at root\Portals\0 -> GoogleAnalytics.config vs GoogleTagManager.config

## Affected version
<!-- 
Please add X in at least one of the boxes as appropriate. In order for an issue to be accepted, a developer needs to be able to reproduce the issue on a currently supported version. If you are looking for a workaround for an issue with an older version, please visit the forums at https://dnncommunity.org/forums
-->
* [ ] 10.00.00 alpha build
* [ X ] 09.08.01 latest supported release

## Affected browser
<!-- 
  Check all that apply, and add more if necessary. As appropriate, please specify the exact version(s) of the browser and operating system.
-->
* [ X ] Chrome
* [ X ] Firefox
* [ ] Safari
* [ ] Internet Explorer 11
* [ ] Microsoft Edge (Classic)
* [ ] Microsoft Edge Chromium

--- a/DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs
+++ b/DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs
@@ -246,7 +246,7 @@ namespace DNN.Connectors.GoogleAnalytics
         /// <returns>The string representing a boolean after the correction.</returns>
         private string HandleCustomBoolean(string value)
         {
-            if (value.Trim().Equals("true", StringComparison.OrdinalIgnoreCase))
+            if ((value ?? string.Empty).Trim().Equals("true", StringComparison.OrdinalIgnoreCase))
             {
                 return "true";
             }

--- a/DNN Platform/Connectors/GoogleTagManager/GoogleTagManagerConnector.cs
+++ b/DNN Platform/Connectors/GoogleTagManager/GoogleTagManagerConnector.cs
@@ -6,7 +6,10 @@ namespace DNN.Connectors.GoogleTagManager
 {
     using System;
     using System.Collections.Generic;
+    using System.Web;
+    using System.Xml;
 
+    using DotNetNuke.Common;
     using DotNetNuke.Entities.Portals;
     using DotNetNuke.Services.Analytics.Config;
     using DotNetNuke.Services.Connections;
@@ -179,6 +182,11 @@ namespace DNN.Connectors.GoogleTagManager
                     });
 
                     AnalyticsConfiguration.SaveConfig("GoogleTagManager", config);
+
+                    if (!isDeactivating)
+                    {
+                        this.EnsureScriptInConfig();
+                    }
                 }
 
                 return isValid;
@@ -187,6 +195,43 @@ namespace DNN.Connectors.GoogleTagManager
             {
                 Exceptions.LogException(ex);
                 return false;
+            }
+        }
+
+        /// <summary>
+        /// Check if there's an AnalyticsEngine element in siteanalytics.config for this connector. If not, adds the default one.
+        /// </summary>
+        private void EnsureScriptInConfig()
+        {
+            var applicationMappath = HttpContext.Current.Server.MapPath("\\");
+            var file = applicationMappath + "\\SiteAnalytics.config";
+            var xdoc = new XmlDocument();
+            xdoc.Load(file);
+            var found = false;
+            foreach (XmlNode engineTypeNode in xdoc.SelectNodes("/AnalyticsEngineConfig/Engines/AnalyticsEngine/EngineType"))
+            {
+                if (engineTypeNode.InnerText.Contains("DotNetNuke.Services.Analytics.GoogleTagManagerEngine"))
+                {
+                    found = true;
+                    break;
+                }
+            }
+
+            if (!found)
+            {
+                var fileGtm = applicationMappath + "\\DesktopModules\\Connectors\\GoogleTagManager\\GoogleTagManager.config";
+                var xdocGtm = new XmlDocument();
+                xdocGtm.Load(fileGtm);
+
+                var enginesElement = xdoc.SelectSingleNode("/AnalyticsEngineConfig/Engines");
+                foreach (XmlNode engineNode in xdocGtm.SelectNodes("/AnalyticsEngineConfig/Engines/AnalyticsEngine"))
+                {
+                    var engineFrag = xdoc.CreateDocumentFragment();
+                    engineFrag.InnerXml = engineNode.OuterXml;
+                    enginesElement.AppendChild(engineFrag);
+                }
+
+                xdoc.Save(file);
             }
         }
 

--- a/DNN Platform/Connectors/GoogleTagManager/GoogleTagManagerConnector.cs
+++ b/DNN Platform/Connectors/GoogleTagManager/GoogleTagManagerConnector.cs
@@ -198,7 +198,7 @@ namespace DNN.Connectors.GoogleTagManager
         /// <returns>The string representing a boolean after the correction.</returns>
         private string HandleCustomBoolean(string value)
         {
-            if (value.Trim().Equals("true", StringComparison.OrdinalIgnoreCase))
+            if ((value ?? string.Empty).Trim().Equals("true", StringComparison.OrdinalIgnoreCase))
             {
                 return "true";
             }

--- a/DNN Platform/Connectors/GoogleTagManager/Scripts/connector.js
+++ b/DNN Platform/Connectors/GoogleTagManager/Scripts/connector.js
@@ -69,7 +69,7 @@ define(["jquery", "knockout", "templatePath/scripts/config", "templatePath/scrip
 
                     // Set the isDeactivating flag to true to override the default save behaviour
                     // Temporary workaround until delete functionality on connectors is improved
-                    conn.configurations[6].value("true");
+                    conn.configurations[2].value("true");
                     wasDeactivated = true;
                     conn.save(conn, e, onSaveComplete.bind(this, conn, conn.id));
                 }
@@ -100,7 +100,7 @@ define(["jquery", "knockout", "templatePath/scripts/config", "templatePath/scrip
 
                         // Set the isDeactivating flag to true to override the default save behaviour
                         // Temporary workaround until delete functionality on connectors is improved
-                        conn.configurations[6].value("true");
+                        conn.configurations[2].value("true");
                         wasDeactivated = true;
                         conn.save(conn, e, onSaveComplete.bind(this, conn, conn.id));
                     }

--- a/DNN Platform/Library/Services/Analytics/Config/AnalyticsConfiguration.cs
+++ b/DNN Platform/Library/Services/Analytics/Config/AnalyticsConfiguration.cs
@@ -84,7 +84,7 @@ namespace DotNetNuke.Services.Analytics.Config
                     {
                         var setting = new AnalyticsSetting();
                         setting.SettingName = nav.SelectSingleNode("SettingName").Value;
-                        setting.SettingValue = nav.SelectSingleNode("SettingValue").Value;
+                        setting.SettingValue = nav.SelectSingleNode("SettingValue")?.Value;
                         Config.Settings.Add(setting);
                     }
 

